modelforge/potential/sake.py

```diff
@@ -0,0 +1,382 @@
+                             (len(idx_i), self.nr_edge_basis * self.nr_heads))
+
+    def forward(
+            self,
+            h: torch.Tensor,
+            x: torch.Tensor,
+            v: torch.Tensor,
+            pairlist: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Compute interaction output.
+
+        Parameters
+        ----------
+        h : torch.Tensor
+            Scalar input values of shape [nr_of_atoms_in_systems, nr_atom_basis].
+        x : torch.Tensor
+            Equivariant position input values of shape [nr_of_atoms_in_systems, geometry_basis].
+        v : torch.Tensor
+            Equivariant velocity input values of shape [nr_of_atoms_in_systems, geometry_basis].
+        pairlist : torch.Tensor, shape (2, nr_pairs)
+        epsilon : float
+            Stability constant to prevent numerical instabilities (default is 1e-8).
+
+        Returns
+        -------
+        Tuple[torch.Tensor, torch.Tensor]
+            Updated scalar and vector representations (q, mu).
+        """
+        idx_i, idx_j = pairlist
+        nr_of_atoms_in_all_systems, _ = x.shape
+        r_ij = x[idx_j] - x[idx_i]
+        d_ij = torch.norm(r_ij, dim=-1)
+        dir_ij = r_ij / (d_ij.unsqueeze(-1) + self.epsilon)
+
+        h_ij_edge = self.edge_model(h[idx_i], h[idx_j], d_ij)
+        h_ij_semantic = self.get_semantic_attention(h_ij_edge, idx_i, d_ij, nr_of_atoms_in_all_systems)
+        h_i_semantic = self.aggregate(h_ij_semantic, idx_i, nr_of_atoms_in_all_systems)
+        combinations = self.get_combinations(h_ij_semantic, dir_ij)
+        h_i_spatial = self.get_spatial_attention(combinations, idx_i, dir_ij, nr_of_atoms_in_all_systems)
+        h_updated = self.update_node(h, h_i_semantic, h_i_spatial)
+        v_updated = self.update_velocity(v, h, combinations, idx_i)
+        x_updated = x + v_updated
+
+        return h_updated, x_updated, v_updated
+
+
+class LightningSAKE(SAKE, LightningModuleMixin):
+    def __init__(
+            self,
+            embedding: nn.Module,
+            nr_interaction_blocks: int,
+            nr_heads: int,
+            radial_basis: nn.Module,
+            cutoff: nn.Module,
+            activation: Optional[Callable] = F.silu,
+            loss: Type[nn.Module] = nn.MSELoss(),
+            optimizer: Type[torch.optim.Optimizer] = torch.optim.Adam,
+            lr: float = 1e-3,
+    ) -> None:
+        """PyTorch Lightning version of the SAKE model."""
+
+        super().__init__(
+            embedding_module=embedding,
+            nr_interaction_blocks=nr_interaction_blocks,
+            nr_heads=nr_heads,
+            radial_basis_module=radial_basis,
+            cutoff_module=cutoff,
+            activation=activation,
+        )
+        self.loss_function = loss
+        self.optimizer = optimizer
+        self.learning_rate = lr
+
+
+class ContinuousFilterConvolutionWithConcatenation(nn.Module):
```

Comment:
Can you please add docstrings explaining the functionality and use case, and the parameters that are passed?